src/cmd/serve.rs

```diff
@@ -116,38 +119,70 @@ async fn serve(
-                trace!("websocket got connection");
-                if let Ok(m) = rx.recv().await {
-                    trace!("notify of reload");
-                    let _ = user_ws_tx.send(m).await;
+    let reload_tx_clone = reload_tx.clone();
+
+    // WebSocket handler for live reload
+    let websocket_handler = move |ws: WebSocketUpgrade| async move {
+        let reload_tx = reload_tx_clone.clone();
+        ws.on_upgrade(move |socket| websocket_connection(socket, reload_tx))
+    };
+
+    // Serve static files with custom 404 fallback
+    let file_404_path = build_dir.join(file_404);
+    let serve_dir = ServeDir::new(&build_dir).not_found_service(tower::service_fn(
+        move |_: axum::http::Request<axum::body::Body>| {
+            let path = file_404_path.clone();
+            async move {
+                match tokio::fs::read(&path).await {
+                    Ok(contents) => {
+                        let mime = mime_guess::from_path(&path).first_or_octet_stream();
+                        Ok::<axum::response::Response, std::convert::Infallible>(
+                            (
+                                axum::http::StatusCode::NOT_FOUND,
+                                [(axum::http::header::CONTENT_TYPE, mime.as_ref())],
+                                contents,
+                            )
+                                .into_response(),
+                        )
+                    }
+                    Err(_) => {
+                        Ok((axum::http::StatusCode::NOT_FOUND, "404 Not Found").into_response())
+                    }
                 }
-            })
-        });
-    // A warp Filter that serves from the filesystem.
-    let book_route = warp::fs::dir(build_dir.clone());
-    // The fallback route for 404 errors
-    let fallback_route = warp::fs::file(build_dir.join(file_404))
-        .map(|reply| warp::reply::with_status(reply, warp::http::StatusCode::NOT_FOUND));
-    let routes = livereload.or(book_route).or(fallback_route);
+            }
+        },
+    ));
+
+    // Build the router
+    let app = Router::new()
+        .route(
+            &format!("/{}", LIVE_RELOAD_ENDPOINT),
+            get(websocket_handler),
+        )
+        .fallback_service(serve_dir);
 
     std::panic::set_hook(Box::new(move |panic_info| {
         // exit if serve panics
         error!("Unable to serve: {}", panic_info);
         std::process::exit(1);
     }));
 
-    warp::serve(routes).run(address).await;
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|_| panic!("Unable to bind to {}", address));
+
+    axum::serve(listener, app).await.unwrap();
+}
+
+async fn websocket_connection(ws: WebSocket, reload_tx: broadcast::Sender<Message>) {
+    let (mut user_ws_tx, _user_ws_rx) = ws.split();
+    let mut rx = reload_tx.subscribe();
+
+    trace!("websocket got connection");
+    while let Ok(m) = rx.recv().await {
```

Comment:
I'm a little unclear why this is a loop. The original code doesn't use a loop, and https://github.com/rust-lang/mdBook/pull/2359 also does not use a loop. Can you say more about that?

Reply:
Just defensive - if the websocket is still open then we might as well forward reload triggers. Reverted.